src/audio/pipeline/pipeline-stream.c

```diff
@@ -268,24 +270,123 @@ static void pipeline_trigger_xrun(struct pipeline *p, struct comp_dev **host)
 	} while (true);
 }
 
+#if CONFIG_KCPS_DYNAMIC_CLOCK_CONTROL
+static int add_pipeline_cps_consumption(struct comp_dev *current,
+					struct comp_buffer *calling_buf,
+					struct pipeline_walk_context *ctx, int dir)
+{
+	struct pipeline_data *ppl_data = ctx->comp_data;
+	struct ipc4_base_module_cfg *cd  = NULL;
+
+	pipe_dbg(ppl_data->p, "add_pipeline_cps_consumption(), current->comp.id = %u, dir = %u",
+		 dev_comp_id(current), dir);
+
+	if (!comp_is_single_pipeline(current, ppl_data->start)) {
+		pipe_dbg(ppl_data->p, "add_pipeline_cps_consumption(), current is from another pipeline");
+		return 0;
+	}
+
+	/* complete component init */
+	current->pipeline = ppl_data->p;
+
+	/* modules created throug module adapter have different priv_data */
+	if (current->drv->type != SOF_COMP_MODULE_ADAPTER) {
+		cd = comp_get_drvdata(current);
+	} else {
+		struct processing_module *mod = comp_get_drvdata(current);
+		struct module_data *md = &mod->priv;
+		cd = &md->cfg.base_cfg;
+	}
+
+	if (cd->cpc == 0) {
+		/* Use maximum clock budget, assume 1ms chunk size */
+		cd->cpc = CLK_MAX_CPU_HZ / 1000;
+		tr_warn(pipe,
+			"0 CPS requested for module: %#x, core: %d using safe max CPC: %u",
+			current->ipc_config.id, ppl_data->p->core, cd->cpc);
+	}
+
+	int kcps = cd->cpc * 1000 / ppl_data->p->period;
+	tr_info(pipe, "Registering KCPS consumption: %d, core: %d", kcps, ppl_data->p->core);
+	core_kcps_adjust(ppl_data->p->core, kcps);
```

Comment:
If none of the module instances have recorded CPC value and the we have 6 components active then the kcps will overflow to negative as the `kcps_consumption` is an `int` array...

Reply:
yeah, thats true. Previously it was intended to fail when 0 cpc received but later it was allowed so we ended in possible negative value. Probably just an int64?

Reply:
@abonislawski @ujfalusi Is this addressed? Does core_kcps_adjust() handle this case and not overflow?

Reply:
@abonislawski ping?

Reply:
**fixed this** but actually freq inside cpu-clk-manager could overflow, not kcps_consumption
also changed add_pipeline_cps_consumption flow so core_kcps_adjust will be called only once for one pipeline